cmd/server/main.go

```diff
@@ -340,6 +342,61 @@ func rateLimitMiddleware(maxReq int, windowLength time.Duration) func(http.Handl
 		}))
 }
 
+func addUserIdentityToContextIfLoggedIn(logger *zap.Logger, db *sqldb.DB) func(http.Handler) http.Handler {
+	fn := func(c context.Context) (context.Context, error) {
+		token, _, err := jwtauth.FromContext(c)
+		if err != nil {
+			return nil, fmt.Errorf("error getting authorization token: %w", err)
+		}
+		if token == nil {
+			return nil, fmt.Errorf("nil authorization token")
+		}
+		emailsClaim, ok := token.PrivateClaims()["emails"]
+		if !ok {
+			return nil, fmt.Errorf("no email claim in token")
+		}
+		emails, ok := emailsClaim.([]interface{})
+		if !ok || len(emails) == 0 {
+			return nil, fmt.Errorf("couldn't find email claim in token: %T", emailsClaim)
+		}
+		// TODO(#18) Handle Multiple Emails in the Token Claims gracefully
+		if len(emails) > 1 {
+			return nil, fmt.Errorf("multiple emails in token: %+v", emails)
+		}
+		email, ok := emails[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("wrong type for email claim: %T", emails[0])
+		}
+		canonical, err := pacta.CanonicalizeEmail(email)
+		if err != nil {
+			return nil, fmt.Errorf("invalid email on token: %q", email)
+		}
+		authnID := token.Subject()
+		if authnID == "" {
+			return nil, fmt.Errorf("couldn't find authn id in jwt")
+		}
+		user, err := db.GetOrCreateUserByAuthn(db.NoTxn(c), pacta.AuthnMechanism_EmailAndPass, authnID, email, canonical)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get user by authn: %w", err)
+		}
+		return jwtauth.WithUserId(c, string(user.ID)), nil
+	}
```

Comment:
Discussed offline:

* Move this into a package/struct, like `session.WithAuthn`
* Use a 'read-only' DB function (the `GetOrCreate` should only be used on login)
* Use `session.{WithUserID,UserIDFromContext}` for setting/retrieving the ID

Reply:
Done.